qiime/differential_abundance.py

```diff
@@ -19,18 +19,43 @@
 __email__ = "sophie.sjw@gmail.com"
 
 
+def check_mapping_file_category(mapping_fp, mapping_category, pmf, subcategory_1, subcategory_2):
+    md_map = MetadataMap.parseMetadataMap(mapping_fp)
+    if not mapping_category in md_map.CategoryNames:
+        raise ValueError("Category '%s' not found in mapping file "
+                         "columns." % mapping_category)
 
+    if md_map.hasSingleCategoryValue(mapping_category):
+        raise ValueError("All values in category '%s' are the "
+                         "same. These statistical methods "
+                         "cannot operate on a category that "
+                         "creates only a single group of samples "
+                         "(e.g. there are no 'between' distances "
+                         "because there is only a single group)."
+                         % (mapping_category))
+
+    all_subcategories = {pmf[k][mapping_category] for k in pmf}
+
+    if subcategory_1 not in all_subcategories:
+        raise ValueError("Subcategory_1 (-x) '%s' not found in selected "
+                         "mapping file column." % subcategory_1)
+
+    if subcategory_2 not in all_subcategories:
+        raise ValueError("Subcategory_2 (-y) '%s' not found in selected "
+                         "mapping file column." % subcategory_2)
+
+                    
 def DA_fitZIG(input_path, out_path, mapping_fp, mapping_category, subcategory_1, subcategory_2):
    """perform metagenomeSeq's Zero Inflated Gaussian (ZIG) OTU differential abundance testing"""
    base_fname, ext = splitext(input_path)
    json_infile = base_fname+'_json.biom'
    tmp_bt = load_table(input_path)
    tmp_pmf, _ = parse_mapping_file_to_dict(mapping_fp)
+   check_mapping_file_category(mapping_fp, mapping_category, tmp_pmf, subcategory_1, subcategory_2)
    tmp_bt.add_metadata(tmp_pmf, 'sample')
    open(str(json_infile),'w').write(tmp_bt.to_json('forR'))
 
    run_fitZIG(json_infile, out_path, mapping_category, subcategory_1, subcategory_2)
-   remove(json_infile)
```

Comment:
I think it makes more sense to create and delete the temporary JSON BIOM table in Python instead of in R. The R scripts could be used directly by an advanced user and we wouldn't want to delete their BIOM table in that case.

Also, I noticed that the temporary files aren't created in a safe manner: suppose the user has an input BIOM table named `foo.biom`, and another table named `foo_json.biom`. The script would overwrite the latter table with the temporary file.

Here's one way to safely handle temporary files in Python, also taking advantage of the user's QIIME temp directory:

```python
import tempfile
from qiime.util import get_qiime_temp_dir
with tempfile.NamedTemporaryFile(dir=get_qiime_temp_dir(),
                                 prefix='QIIME-differential-abundance-temp-table-',
                                 suffix='.biom') as temp_fh:
    temp_fh.write(tmp_bt.to_json('forR'))
    temp_fh.flush()
    run_fitZIG(temp_fh.name, out_path, mapping_category, subcategory_1, subcategory_2)
```

This will safely create a temporary file in the user's QIIME temp directory, which is then written to, and then the filepath is passed to `run_fitZIG`. Whenever execution leaves the `with` statement, the temporary file is automatically closed and deleted (regardless of whether there was an error or not).

Can you please make this change throughout this code and normalize_table.py (if applicable)?